src/python/pants/backend/codegen/protobuf/java/rules.py

```diff
@@ -0,0 +1,124 @@
+    input = ProtobufSourceField
+    output = PythonSourceField
+
+
+@rule(desc="Generate Java from Protobuf", level=LogLevel.DEBUG)
+async def generate_java_from_protobuf(
+    request: GenerateJavaFromProtobufRequest,
+    protoc: Protoc,
+) -> GeneratedSources:
+    download_protoc_request = Get(
+        DownloadedExternalTool, ExternalToolRequest, protoc.get_request(Platform.current)
+    )
+
+    output_dir = "_generated_files"
+    create_output_dir_request = Get(Digest, CreateDigest([Directory(output_dir)]))
+
+    # Protoc needs all transitive dependencies on `protobuf_libraries` to work properly. It won't
+    # actually generate those dependencies; it only needs to look at their .proto files to work
+    # with imports.
+    transitive_targets = await Get(
+        TransitiveTargets, TransitiveTargetsRequest([request.protocol_target.address])
+    )
+
+    # NB: By stripping the source roots, we avoid having to set the value `--proto_path`
+    # for Protobuf imports to be discoverable.
+    all_stripped_sources_request = Get(
+        StrippedSourceFiles,
+        SourceFilesRequest(
+            tgt[ProtobufSourceField]
+            for tgt in transitive_targets.closure
+            if tgt.has_field(ProtobufSourceField)
+        ),
+    )
+    target_stripped_sources_request = Get(
+        StrippedSourceFiles, SourceFilesRequest([request.protocol_target[ProtobufSourceField]])
+    )
+
+    (
+        downloaded_protoc_binary,
+        empty_output_dir,
+        all_sources_stripped,
+        target_sources_stripped,
+    ) = await MultiGet(
+        download_protoc_request,
+        create_output_dir_request,
+        all_stripped_sources_request,
+        target_stripped_sources_request,
+    )
+
+    unmerged_digests = [
+        all_sources_stripped.snapshot.digest,
+        downloaded_protoc_binary.digest,
+        empty_output_dir,
+    ]
+    input_digest = await Get(Digest, MergeDigests(unmerged_digests))
+
+    argv = [downloaded_protoc_binary.exe, "--java_out", output_dir]
+
+    argv.extend(target_sources_stripped.snapshot.files)
+    result = await Get(
+        ProcessResult,
+        Process(
+            argv,
+            input_digest=input_digest,
+            description=f"Generating Java sources from {request.protocol_target.address}.",
+            level=LogLevel.DEBUG,
+            output_directories=(output_dir,),
+        ),
+    )
+
+    normalized_digest, source_root = await MultiGet(
+        Get(Digest, RemovePrefix(result.output_digest, output_dir)),
+        Get(SourceRoot, SourceRootRequest, SourceRootRequest.for_target(request.protocol_target)),
+    )
+
+    source_root_restored = (
+        await Get(Snapshot, AddPrefix(normalized_digest, source_root.path))
+        if source_root.path != "."
+        else await Get(Snapshot, Digest, normalized_digest)
+    )
+    return GeneratedSources(source_root_restored)
+
+
+def rules():
+    return [
+        *collect_rules(),
+        *pex.rules(),
```

Comment:
Was this necessary? Or bad copy pasta?

Reply:
I get this rule graph error with it commented out:

```
E       ValueError: Encountered 12 rule graph errors:
E         No installed rules return the type DownloadedExternalTool to satisfy Get(DownloadedExternalTool, ExternalToolRequest) for @rule(pants.backend.codegen.protobuf.java.rules:39:generate_java_from_protobuf(GenerateJavaFromProtobufRequest, Protoc) -> GeneratedSources, gets=[Get(DownloadedExternalTool, ExternalToolRequest), Get(PyDigest, CreateDigest), Get(TransitiveTargets, TransitiveTargetsRequest), Get(StrippedSourceFiles, SourceFilesRequest), Get(PyDigest, PyMergeDigests), Get(ProcessResult, Process), Get(PyDigest, PyRemovePrefix), Get(SourceRoot, SourceRootRequest), Get(PySnapshot, PyAddPrefix), Get(PySnapshot, PyDigest)]).
E           Ensure that the rule you are expecting to use is installed.
E         No installed rules return the type GenerateJavaFromProtobufRequest, and it was not provided by potential callers of @rule(pants.backend.codegen.protobuf.java.rules:39:generate_java_from_protobuf(GenerateJavaFromProtobufRequest, Protoc) -> GeneratedSources, gets=[Get(DownloadedExternalTool, ExternalToolRequest), Get(PyDigest, CreateDigest), Get(TransitiveTargets, TransitiveTargetsRequest), Get(StrippedSourceFiles, SourceFilesRequest), Get(PyDigest, PyMergeDigests), Get(ProcessResult, Process), Get(PyDigest, PyRemovePrefix), Get(SourceRoot, SourceRootRequest), Get(PySnapshot, PyAddPrefix), Get(PySnapshot, PyDigest)]).
E           If that type should be computed by a rule, ensure that that rule is installed.
E           If it should be provided by a caller, ensure that it is included in any relevant Query or Get.
E         No source of dependency Get(ProcessResult, Process) for @rule(pants.backend.codegen.protobuf.java.rules:39:generate_java_from_protobuf(GenerateJavaFromProtobufRequest, Protoc) -> GeneratedSources, gets=[Get(DownloadedExternalTool, ExternalToolRequest), Get(PyDigest, CreateDigest), Get(TransitiveTargets, TransitiveTargetsRequest), Get(StrippedSourceFiles, SourceFilesRequest), Get(PyDigest, PyMergeDigests), Get(ProcessResult, Process), Get(PyDigest, PyRemovePrefix), Get(SourceRoot, SourceRootRequest), Get(PySnapshot, PyAddPrefix), Get(PySnapshot, PyDigest)]). All potential sources were eliminated: []
E         No source of dependency Get(PyDigest, CreateDigest) for @rule(pants.backend.codegen.protobuf.java.rules:39:generate_java_from_protobuf(GenerateJavaFromProtobufRequest, Protoc) -> GeneratedSources, gets=[Get(DownloadedExternalTool, ExternalToolRequest), Get(PyDigest, CreateDigest), Get(TransitiveTargets, TransitiveTargetsRequest), Get(StrippedSourceFiles, SourceFilesRequest), Get(PyDigest, PyMergeDigests), Get(ProcessResult, Process), Get(PyDigest, PyRemovePrefix), Get(SourceRoot, SourceRootRequest), Get(PySnapshot, PyAddPrefix), Get(PySnapshot, PyDigest)]). All potential sources were eliminated: []
E         No source of dependency Get(PyDigest, PyMergeDigests) for @rule(pants.backend.codegen.protobuf.java.rules:39:generate_java_from_protobuf(GenerateJavaFromProtobufRequest, Protoc) -> GeneratedSources, gets=[Get(DownloadedExternalTool, ExternalToolRequest), Get(PyDigest, CreateDigest), Get(TransitiveTargets, TransitiveTargetsRequest), Get(StrippedSourceFiles, SourceFilesRequest), Get(PyDigest, PyMergeDigests), Get(ProcessResult, Process), Get(PyDigest, PyRemovePrefix), Get(SourceRoot, SourceRootRequest), Get(PySnapshot, PyAddPrefix), Get(PySnapshot, PyDigest)]). All potential sources were eliminated: []
E         No source of dependency Get(PyDigest, PyRemovePrefix) for @rule(pants.backend.codegen.protobuf.java.rules:39:generate_java_from_protobuf(GenerateJavaFromProtobufRequest, Protoc) -> GeneratedSources, gets=[Get(DownloadedExternalTool, ExternalToolRequest), Get(PyDigest, CreateDigest), Get(TransitiveTargets, TransitiveTargetsRequest), Get(StrippedSourceFiles, SourceFilesRequest), Get(PyDigest, PyMergeDigests), Get(ProcessResult, Process), Get(PyDigest, PyRemovePrefix), Get(SourceRoot, SourceRootRequest), Get(PySnapshot, PyAddPrefix), Get(PySnapshot, PyDigest)]). All potential sources were eliminated: []
E         No source of dependency Get(PySnapshot, PyAddPrefix) for @rule(pants.backend.codegen.protobuf.java.rules:39:generate_java_from_protobuf(GenerateJavaFromProtobufRequest, Protoc) -> GeneratedSources, gets=[Get(DownloadedExternalTool, ExternalToolRequest), Get(PyDigest, CreateDigest), Get(TransitiveTargets, TransitiveTargetsRequest), Get(StrippedSourceFiles, SourceFilesRequest), Get(PyDigest, PyMergeDigests), Get(ProcessResult, Process), Get(PyDigest, PyRemovePrefix), Get(SourceRoot, SourceRootRequest), Get(PySnapshot, PyAddPrefix), Get(PySnapshot, PyDigest)]). All potential sources were eliminated: []
E         No source of dependency Get(PySnapshot, PyDigest) for @rule(pants.backend.codegen.protobuf.java.rules:39:generate_java_from_protobuf(GenerateJavaFromProtobufRequest, Protoc) -> GeneratedSources, gets=[Get(DownloadedExternalTool, ExternalToolRequest), Get(PyDigest, CreateDigest), Get(TransitiveTargets, TransitiveTargetsRequest), Get(StrippedSourceFiles, SourceFilesRequest), Get(PyDigest, PyMergeDigests), Get(ProcessResult, Process), Get(PyDigest, PyRemovePrefix), Get(SourceRoot, SourceRootRequest), Get(PySnapshot, PyAddPrefix), Get(PySnapshot, PyDigest)]). All potential sources were eliminated: []
E         No source of dependency Get(SourceRoot, SourceRootRequest) for @rule(pants.backend.codegen.protobuf.java.rules:39:generate_java_from_protobuf(GenerateJavaFromProtobufRequest, Protoc) -> GeneratedSources, gets=[Get(DownloadedExternalTool, ExternalToolRequest), Get(PyDigest, CreateDigest), Get(TransitiveTargets, TransitiveTargetsRequest), Get(StrippedSourceFiles, SourceFilesRequest), Get(PyDigest, PyMergeDigests), Get(ProcessResult, Process), Get(PyDigest, PyRemovePrefix), Get(SourceRoot, SourceRootRequest), Get(PySnapshot, PyAddPrefix), Get(PySnapshot, PyDigest)]). All potential sources were eliminated: []
E         No source of dependency Get(StrippedSourceFiles, SourceFilesRequest) for @rule(pants.backend.codegen.protobuf.java.rules:39:generate_java_from_protobuf(GenerateJavaFromProtobufRequest, Protoc) -> GeneratedSources, gets=[Get(DownloadedExternalTool, ExternalToolRequest), Get(PyDigest, CreateDigest), Get(TransitiveTargets, TransitiveTargetsRequest), Get(StrippedSourceFiles, SourceFilesRequest), Get(PyDigest, PyMergeDigests), Get(ProcessResult, Process), Get(PyDigest, PyRemovePrefix), Get(SourceRoot, SourceRootRequest), Get(PySnapshot, PyAddPrefix), Get(PySnapshot, PyDigest)]). All potential sources were eliminated: []
E         No source of dependency Get(TransitiveTargets, TransitiveTargetsRequest) for @rule(pants.backend.codegen.protobuf.java.rules:39:generate_java_from_protobuf(GenerateJavaFromProtobufRequest, Protoc) -> GeneratedSources, gets=[Get(DownloadedExternalTool, ExternalToolRequest), Get(PyDigest, CreateDigest), Get(TransitiveTargets, TransitiveTargetsRequest), Get(StrippedSourceFiles, SourceFilesRequest), Get(PyDigest, PyMergeDigests), Get(ProcessResult, Process), Get(PyDigest, PyRemovePrefix), Get(SourceRoot, SourceRootRequest), Get(PySnapshot, PyAddPrefix), Get(PySnapshot, PyDigest)]). All potential sources were eliminated: []
E         No source of dependency Protoc for @rule(pants.backend.codegen.protobuf.java.rules:39:generate_java_from_protobuf(GenerateJavaFromProtobufRequest, Protoc) -> GeneratedSources, gets=[Get(DownloadedExternalTool, ExternalToolRequest), Get(PyDigest, CreateDigest), Get(TransitiveTargets, TransitiveTargetsRequest), Get(StrippedSourceFiles, SourceFilesRequest), Get(PyDigest, PyMergeDigests), Get(ProcessResult, Process), Get(PyDigest, PyRemovePrefix), Get(SourceRoot, SourceRootRequest), Get(PySnapshot, PyAddPrefix), Get(PySnapshot, PyDigest)]). All potential sources were eliminated: []
```